core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -213,36 +214,30 @@ class TransactionStateManager(brokerId: Int,
                                              createdTxnMetadata: Option[TransactionMetadata]): Either[Errors, Option[CoordinatorEpochAndTxnMetadata]] = {
     inReadLock(stateLock) {
       val partitionId = partitionFor(transactionalId)
-
       if (loadingPartitions.exists(_.txnPartitionId == partitionId))
-        return Left(Errors.COORDINATOR_LOAD_IN_PROGRESS)
-
-      if (leavingPartitions.exists(_.txnPartitionId == partitionId))
-        Right(Errors.NOT_COORDINATOR)
-
-      transactionMetadataCache.get(partitionId) match {
-        case Some(cacheEntry) =>
-          cacheEntry.metadataPerTransactionalId.get(transactionalId) match {
-            case null =>
-              createdTxnMetadata match {
-                case None =>
-                  Right(None)
-
-                case Some(txnMetadata) =>
+        Left(Errors.COORDINATOR_LOAD_IN_PROGRESS)
+      else if (leavingPartitions.exists(_.txnPartitionId == partitionId))
+        Left(Errors.NOT_COORDINATOR)
+      else {
+        transactionMetadataCache.get(partitionId) match {
+          case Some(cacheEntry) =>
+            Right(cacheEntry.metadataPerTransactionalId.get(transactionalId) match {
+              case null =>
+                createdTxnMetadata.map { txnMetadata =>
                   val currentTxnMetadata = cacheEntry.metadataPerTransactionalId.putIfNotExists(transactionalId, txnMetadata)
                   if (currentTxnMetadata != null) {
-                    Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata)))
+                    CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata)
                   } else {
-                    Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata)))
+                    CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata)
                   }
-              }
-
-            case currentTxnMetadata =>
-              Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata)))
-          }
+                }
+              case currentTxnMetadata =>
+                Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata))
```

Comment:
Would it work to move the creation of this instance higher up and to just do the txn metadata within these two `case` statements?